File: websocket-public-chat/profile-a/zig/src/protocol.zig
const std = @import("std");

pub const msg_join = "join";
pub const msg_chat = "chat";
pub const msg_ping = "ping";
pub const msg_pong = "pong";
pub const msg_leave = "leave";

pub const chat_payload_size: usize = 128;
pub const room = "public";
pub const rate_limit_msg_per_sec: usize = 10;
pub const ping_interval_sec: u64 = 30;

pub const Message = struct {
    type: []const u8,
    room: ?[]const u8 = null,
    user: ?[]const u8 = null,
    text: ?[]const u8 = null,
    ts: ?i64 = null,
};

test "test_parse_json_message" {
    const allocator = std.testing.allocator;
    const payload =
        \\{"type":"chat","room":"public","user":"client-01","text":"hello","ts":1700000000}
    ;

    var parsed = try std.json.parseFromSlice(Message, allocator, payload, .{});
    defer parsed.deinit();

    try std.testing.expectEqualStrings(msg_chat, parsed.value.type);
    try std.testing.expect(parsed.value.user != null);
    try std.testing.expectEqualStrings("client-01", parsed.value.user.?);
}

File: websocket-public-chat/profile-a/go/protocol.go
package main

import (
	"encoding/json"
	"strings"
)

const (
	MsgJoin  = "join"
	MsgChat  = "chat"
	MsgPing  = "ping"
	MsgPong  = "pong"
	MsgLeave = "leave"
)

const (
	ChatPayloadSize    = 128
	Room               = "public"
	RateLimitMsgPerSec = 10
	PingIntervalSec    = 30
)

type Message struct {
	Type string `json:"type"`
	Room string `json:"room,omitempty"`
	User string `json:"user,omitempty"`
	Text string `json:"text,omitempty"`
	Ts   int64  `json:"ts,omitempty"`
}

func padToSize(text string, size int) string {
	if len(text) >= size {
		return text[:size]
	}
	return text + strings.Repeat(" ", size-len(text))
}

func marshalChatMessage(user, text string, ts int64) ([]byte, error) {
	msg := Message{
		Type: MsgChat,
		Room: Room,
		User: user,
		Text: text,
		Ts:   ts,
	}

	for {
		data, err := json.Marshal(msg)
		if err != nil {
			return nil, err
		}

		if len(data) == ChatPayloadSize {
			return data, nil
		}

		if len(data) < ChatPayloadSize {
			msg.Text += strings.Repeat(" ", ChatPayloadSize-len(data))
			continue
		}

		over := len(data) - ChatPayloadSize
		if over >= len(msg.Text) {
			msg.Text = ""
		} else {
			msg.Text = msg.Text[:len(msg.Text)-over]
		}
	}
}
